src/Credentials/CredentialProvider.php

```diff
@@ -225,15 +236,40 @@ public static function ini($profile = null, $filename = null)
                     . "'$profile' ($filename)");
             }
 
-            return Promise\promise_for(
-                new Credentials(
-                    $data[$profile]['aws_access_key_id'],
-                    $data[$profile]['aws_secret_access_key'],
-                    isset($data[$profile]['aws_security_token'])
-                        ? $data[$profile]['aws_security_token']
-                        : null
-                )
+            $credentials = new Credentials(
+                $data[$profile]['aws_access_key_id'],
+                $data[$profile]['aws_secret_access_key'],
+                isset($data[$profile]['aws_security_token'])
+                    ? $data[$profile]['aws_security_token']
+                    : null
             );
+
+            self::tryToCache($credentials, $config);
```

Comment:
This shouldn't know about the cache either. Caching should wrap other providers like the memoize function.